sql/opt: panic "unsupported op subquery"

This panic was found with sqlsmith random query generation. #21852

```
root@:26257/tpcc> create table a (a int primary key);
CREATE TABLE

Time: 5.923276ms

root@:26257/tpcc> select * from a where cast(coalesce((select 'a'), (select 'a')) as bytea) < 'a';
driver: bad connection
warning: connection lost!
opening new connection: all session settings will be lost
```

```
E180323 03:46:17.642130 420 sql/conn_executor.go:514  [n1,client=127.0.0.1:62148,user=root] a SQL panic has occurred while executing "SELECT * FROM a WHERE CAST(COALESCE((SELECT 'a'), (SELECT 'a')) AS BYTEA) < '\\x61'": unsupported op subquery
I180323 03:46:17.642159 420 sql/conn_executor.go:1816  [n1,client=127.0.0.1:62148,user=root] finishing connExecutor
E180323 03:46:17.642286 420 util/log/crash_reporting.go:185  [n1,client=127.0.0.1:62148,user=root] a panic has occurred!
E180323 03:46:17.644499 420 util/log/crash_reporting.go:451  [n1,client=127.0.0.1:62148,user=root] Reported as error 68709f45ac10406f9886e31f83e33f77
panic: unsupported op subquery [recovered]
        panic: panic while executing 1 statements: SELECT * FROM _ WHERE CAST(COALESCE((SELECT _), (SELECT _)) AS BYTEA) < _; caused by unsupported op subquery

goroutine 420 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc4255ea000, 0x606de00, 0xc4255bea00, 0x5a2e1e0, 0xc426a7cd80)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:528 +0x365
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func2(0xc4255ea000, 0x606de00, 0xc4255bea00)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:489 +0x61
panic(0x5a2e1e0, 0xc426a7cd80)
        /usr/local/Cellar/go/1.10/libexec/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc426f818c0, 0xc4261a1780, 0xc4269f37a0, 0x1b0000000a, 0x1b0000000a, 0x2)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:78 +0x162
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildCoalesce(0xc426f818c0, 0xc4261a1780, 0xc4269f37a0, 0x570000000d, 0xc42561fde8, 0x503d325)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:230 +0x113
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc426f818c0, 0xc4261a1780, 0xc4269f37a0, 0x570000000d, 0x570000000d, 0xc42561ff90)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:76 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildCast(0xc426f818c0, 0xc4261a1780, 0xc4269f37a0, 0x530000000e, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:218 +0x92
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc426f818c0, 0xc4261a1780, 0xc4269f37a0, 0x530000000e, 0x530000000e, 0x40)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:76 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildComparison(0xc426f818c0, 0xc4261a1780, 0xc4269f37a0, 0x2c00000010, 0xc42561ffd0, 0x4014699)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:145 +0x92
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc426f818c0, 0xc4261a1780, 0xc4269f37a0, 0x2c00000010, 0x8041d90, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:76 +0x88
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).BuildScalar(0xc426f818c0, 0xc426160d28, 0xc425620308, 0xc4269f37a0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:64 +0x12c
github.com/cockroachdb/cockroach/pkg/sql.(*planner).selectIndex(0xc4255ea3e8, 0x606dec0, 0xc426fd0030, 0xc426160c00, 0x0, 0xc426a88200, 0xc426a882a0, 0x6070201, 0xc426a882a0, 0x5bc5701)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt_index_selection.go:261 +0xf73
github.com/cockroachdb/cockroach/pkg/sql.expandScanNode(0x606dec0, 0xc426fd0030, 0xc4255ea3e8, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0xc426160c00, 0xc426f86420, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:433 +0x10e
github.com/cockroachdb/cockroach/pkg/sql.doExpandPlan(0x606dec0, 0xc426fd0030, 0xc4255ea3e8, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0x6064500, 0xc426160c00, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:271 +0x1dab
github.com/cockroachdb/cockroach/pkg/sql.expandRenderNode(0x606dec0, 0xc426fd0030, 0xc4255ea3e8, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0xc426aa4840, 0xc426160c00, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:446 +0x103
github.com/cockroachdb/cockroach/pkg/sql.doExpandPlan(0x606dec0, 0xc426fd0030, 0xc4255ea3e8, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0x6064400, 0xc426aa4840, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:274 +0x18d6
github.com/cockroachdb/cockroach/pkg/sql.(*planner).expandPlan(0xc4255ea3e8, 0x606dec0, 0xc426fd0030, 0x6064400, 0xc426aa4840, 0x6064400, 0xc426aa4840, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:34 +0xce
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optimizePlan(0xc4255ea3e8, 0x606dec0, 0xc426fd0030, 0x6064400, 0xc426aa4840, 0xc42693dbca, 0x1, 0x1, 0x6064400, 0xc426aa4840, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/optimize.go:47 +0xeb
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc4255ea3e8, 0x606dec0, 0xc426fd0030, 0x6070700, 0xc426a5d080, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:323 +0x2ea
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc4255ea000, 0x606dec0, 0xc426fd0030, 0x6070700, 0xc426a5d080, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:552 +0xb1b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc4255ea000, 0x606dec0, 0xc426fd0030, 0x6070700, 0xc426a5d080, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:369 +0xa8c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc4255ea000, 0x606dec0, 0xc426fd0030, 0x6070700, 0xc426a5d080, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:84 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc4255ea000, 0x606de00, 0xc4255bea00, 0x0, 0x0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:911 +0x1d2b
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc4205081c0, 0x606de00, 0xc4255bea00, 0x0, 0x0, 0xc4255e5061, 0x4, 0xc4255e5052, 0x9, 0x6057340, ...)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:491 +0xf80
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc4205081c0, 0x606de00, 0xc4255bea00, 0xc424ca0380, 0x5400, 0x15000, 0xc420b72190, 0xc420452810, 0xc4255d8570, 0xc4204527f0)
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:260 +0xfc
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:259 +0xf14
```

--- a/pkg/sql/logictest/testdata/logic_test/subquery
+++ b/pkg/sql/logictest/testdata/logic_test/subquery
@@ -512,3 +512,16 @@ root                 ·          ·
            └── scan  ·          ·
 ·                    table      tab4@primary
 ·                    spans      ALL
+
+statement ok
+CREATE TABLE a (a INT PRIMARY KEY)
+
+# Regression test for #24171.
+query I
+SELECT * FROM  a WHERE CAST(COALESCE((SELECT 'a' FROM crdb_internal.zones LIMIT 1 OFFSET 5), (SELECT 'b' FROM pg_catalog.pg_trigger)) AS BYTEA) <= 'a'
+----
+
+# Regression test for #24170.
+query I
+SELECT * FROM a WHERE CAST(COALESCE((SELECT 'a'), (SELECT 'a')) AS bytea) < 'a'
+----

--- a/pkg/sql/opt/optbuilder/scope.go
+++ b/pkg/sql/opt/optbuilder/scope.go
@@ -484,6 +484,11 @@ func (s *scope) VisitPre(expr tree.Expr) (recurse bool, newExpr tree.Expr) {
 		}
 
 	case *tree.ArrayFlatten:
+		if s.builder.AllowUnsupportedExpr {
+			// TODO(rytaft): Temporary fix for #24171 and #24170.
+			break
+		}
+
 		// TODO(peter): the ARRAY flatten operator requires a single column from
 		// the subquery.
 		if sub, ok := t.Subquery.(*tree.Subquery); ok {
@@ -491,6 +496,11 @@ func (s *scope) VisitPre(expr tree.Expr) (recurse bool, newExpr tree.Expr) {
 		}
 
 	case *tree.ComparisonExpr:
+		if s.builder.AllowUnsupportedExpr {
+			// TODO(rytaft): Temporary fix for #24171 and #24170.
+			break
+		}
+
 		switch t.Operator {
 		case tree.In, tree.NotIn, tree.Any, tree.Some, tree.All:
 			if sub, ok := t.Right.(*tree.Subquery); ok {
@@ -499,6 +509,11 @@ func (s *scope) VisitPre(expr tree.Expr) (recurse bool, newExpr tree.Expr) {
 		}
 
 	case *tree.Subquery:
+		if s.builder.AllowUnsupportedExpr {
+			// TODO(rytaft): Temporary fix for #24171 and #24170.
+			break
+		}
+
 		if t.Exists {
 			expr = s.replaceSubquery(t, true /* multi-row */, -1 /* desired-columns */)
 		} else {

